zerver/lib/user_topics.py

```diff
@@ -8,26 +8,68 @@
 
 from zerver.lib.timestamp import datetime_to_timestamp
 from zerver.lib.topic import topic_match_sa
+from zerver.lib.types import UserTopicDict
 from zerver.models import UserProfile, UserTopic, get_stream
 
 
-def get_topic_mutes(
-    user_profile: UserProfile, include_deactivated: bool = False
-) -> List[Tuple[str, str, float]]:
-    query = UserTopic.objects.filter(user_profile=user_profile, visibility_policy=UserTopic.MUTED)
-    # Exclude muted topics that are part of deactivated streams unless
+def get_user_topics(
+    user_profile: UserProfile,
+    include_deactivated: bool = False,
+    include_stream_name: bool = False,
+    visibility_policy: Optional[int] = None,
+) -> List[UserTopicDict]:
+    """
+    Fetches UserTopic objects associated with the target user.
+    * include_deactivated: Whether to include those associated with
+      deactivated streams.
+    * include_stream_name: Whether to include stream names in the
+      returned dictionaries.
+    * visibility_policy: If specified, returns only UserTopic objects
+      with the specified visibility_policy value.
+    """
+    query = UserTopic.objects.filter(user_profile=user_profile)
+
+    if visibility_policy is not None:
+        query = query.filter(visibility_policy=visibility_policy)
+
+    # Exclude user topics that are part of deactivated streams unless
     # explicitly requested.
     if not include_deactivated:
         query = query.filter(stream__deactivated=False)
+
     rows = query.values(
-        "stream__name",
-        "topic_name",
-        "last_updated",
+        "stream_id", "stream__name", "topic_name", "last_updated", "visibility_policy"
+    )
+
+    result = []
+    for row in rows:
+        if not include_stream_name:
+            del row["stream__name"]
+
+        row["last_updated"] = datetime_to_timestamp(row["last_updated"])
+        result.append(row)
+
+    return result
+
+
+def get_topic_mutes(
+    user_profile: UserProfile, include_deactivated: bool = False
+) -> List[Tuple[str, str, int]]:
+
+    user_topics = get_user_topics(
+        user_profile=user_profile,
+        include_deactivated=include_deactivated,
+        include_stream_name=True,
+        visibility_policy=UserTopic.MUTED,
     )
-    return [
-        (row["stream__name"], row["topic_name"], datetime_to_timestamp(row["last_updated"]))
-        for row in rows
-    ]
+
+    result = []
+    for user_topic in user_topics:
+        result.append(
+            (user_topic["stream__name"], user_topic["topic_name"], user_topic["last_updated"])
+        )
+
+    return result
```

Comment:
You can do something more pythonic like the line you're replacing -- list comprehensions are really popular in Python! Whenever you do something with a for loop to generate a new list of the same size, consider a list comprehension like this :)

```
return [
    (user_topic["stream__name"], user_topic["topic_name"], user_topic["last_updated"])
    for user_topic in user_topics
)
```